packages/core/typescript/TypeScriptFormattingService.ts

```diff
@@ -0,0 +1,217 @@
+import * as ts from 'typescript';
+import { IFormattingService, IFormatSettings } from '../types';
+import { IFileSystem } from '../types/FileSystem';
+import { Util } from '../util';
+
+const NEW_LINE_PLACEHOLDER = '//I keep the new line';
+export class FormattingService implements IFormattingService {
+  private _printer: ts.Printer | undefined;
+  private _formatSettingsFromConfig: IFormatSettings = {};
+  private _defaultFormatSettings: IFormatSettings = {
+    indentSize: 3,
+    tabSize: 4,
+    newLineCharacter: ts.sys.newLine,
+    convertTabsToSpaces: true,
+    indentStyle: ts.IndentStyle.Smart,
+    insertSpaceAfterCommaDelimiter: true,
+    insertSpaceAfterSemicolonInForStatements: true,
+    insertSpaceBeforeAndAfterBinaryOperators: true,
+    insertSpaceAfterKeywordsInControlFlowStatements: true,
+    insertSpaceAfterTypeAssertion: true,
+    singleQuotes: true,
+  };
+
+  /**
+   * The printer instance to use to get the source code from the AST.
+   */
+  private get printer(): ts.Printer {
+    if (!this._printer) {
+      this._printer = ts.createPrinter(this.printerOptions);
+    }
+
+    return this._printer;
+  }
```

Comment:
That's weird the service should need that

Reply:
If the service does not operate on a `ts.SourceFIle` but reads the content from the FS, then yes it does not really need a printer.